sw/device/lib/dif/dif_aes.h

```diff
@@ -0,0 +1,359 @@
+ * (automatic or manual).
+ * [TODO - Open question Nr. 13 in #2382]
+ *
+ * @param aes AES state data.
+ * @param key AES Initial Key.
+ * @return `dif_aes_key_set_result_t`.
+ */
+dif_aes_key_set_result_t dif_aes_key_set(const dif_aes_t *aes,
+                                         const uint8_t *key);
+
+/**
+ * AES Initialisation Vector set status codes.
+ *
+ * TODO.
+ */
+typedef enum dif_aes_iv_set_result {
+  kDifAesIvSetOk = kDifAesOk,
+  kDifAesIvSetError = kDifAesError,
+  kDifAesIvSetBadArg = kDifAesBadArg,
+  kDifAesIvSetBusy, /**< Device is in operation, and cannot perform the
+                          requested operation. */
+  kDifAesIvIgnored, /**< Device operates in the ECB Cipher mode, IV is
+                         irrelevant. */
+} dif_aes_iv_set_result_t;
+
+/**
+ * Sets Initialisation Vector (IV).
+ *
+ * The peripheral must be in IDLE state for this operation to take effect. In
+ * CBC and CTR Cipher modes, these registers must be written at least once when
+ * operating in automatic mode.
+ * [TODO - is it different for the manual mode? Open question Nr. 13 in #2382]
+ * The meaning of the Initialisation Vector (IV) depends on the AES Cipher mode:
+ *  - ECB: always ignored.
+ *  - CBC: Initialisation Vector.
+ *  - CTR: Initial Counter Value.
+ *
+ * @param aes AES state data.
+ * @param iv AES Initialisation Vector/Initial Counter Value.
+ * @return `dif_aes_iv_set_result_t`.
+ */
+dif_aes_iv_set_result_t dif_aes_iv_set(const dif_aes_t *aes, const uint8_t *iv);
+
+/**
+ * AES Input Data set status codes.
+ */
+typedef enum dif_aes_in_data_set_result {
+  kDifAesInDataSetOk = kDifAesOk,
+  kDifAesInDataSetError = kDifAesError,
+  kDifAesInDataSetBadArg = kDifAesBadArg,
+  kDifAesInDataSetBusy, /**< Device is in operation, and cannot perform the
+                             requested operation. */
+} dif_aes_in_data_set_result_t;
+
+/**
+ * Sets AES Input Data.
+ *
+ * For maximal performance the `data` must be aligned.
+ *
+ * The behaviour of the AES peripheral is different in automatic and manual
+ * operation mode:
+ *   - automatic: stall condition can occur, which prevents overwriting the
+ *                output data of the previous operation if it had not been read.
+ *
+ *   - manual: peripheral never stalls, and will overwrite the output data
+ *             of the previous operation if it had not been read.
+ *
+ * The input registers must be written at least once when operating in automatic
+ * mode.
+ * [TODO - Don't need to be written in manual? Open question Nr. 13 in #2382]
+ * The meaning of the Input Data depends on the AES operation:
+ *   - Plain text data (encrypt).
+ *   - Cipher text data (decrypt).
+ *
+ * @param aes AES state data.
+ * @param data AES Input Data.
+ * @return `dif_aes_in_data_set_result_t`.
+ */
+dif_aes_in_data_set_result_t dif_aes_in_data_set(const dif_aes_t *aes,
+                                                 const void *data);
```

Comment:
Sure, but wouldn't it place additional requirements on the caller - forcing the buffers that they feed to the function to be of `uint32_t` type to prevent aliasing issues?

If it's `void*` and internally we use `memcpy`, then compiler should be able to figure out when the buffer is aligned and do the copy efficiently?

Reply:
We decided it was better to reflect that AES is handling 4 32-bit integers, rather than to just have this parameter be a buffer. 

One of the major things we didn't like was that `void*` has an implicit length, something the struct makes much more explicit, which we think is a lot safer.

Reply:
`char *example_message[19] = "123412341234123412" // 19 bytes with \0`
How would you propose to feed the handle `example_message` into `somefunc`, given:
`typedef struct somestruct { uint32_t someval[4]; } somestruct_t;`
`void somefunc(somestruct_t input_data);`

Reply:
Done.